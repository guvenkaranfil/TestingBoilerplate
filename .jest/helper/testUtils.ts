import {NavigationContainer} from '@react-navigation/native';
import React from 'react';
import {
  RenderOptions,
  render as rtlRender,
} from '@testing-library/react-native';
import AllProviders from './AllProviders';

function customRender(ui: React.ReactElement, options?: RenderOptions) {
  return rtlRender(ui, {wrapper: AllProviders, ...options});
}

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
