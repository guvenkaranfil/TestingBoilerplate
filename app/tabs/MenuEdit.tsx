import {StyleSheet, Text, View} from 'react-native';
import React from 'react';

export default function MenuEdit() {
  return (
    <View>
      <Text>MenuEdit</Text>
    </View>
  );
}

const styles = StyleSheet.create({});
