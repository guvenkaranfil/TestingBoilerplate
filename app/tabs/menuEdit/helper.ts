export const getCurrentDay = (): number => {
  return Math.random();
};
