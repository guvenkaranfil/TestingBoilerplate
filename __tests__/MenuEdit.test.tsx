import React from 'react';
import {render, screen, userEvent} from '../.jest/helper/testUtils';

import MenuEdit, {IMenuItem} from '../app/tabs/menuEdit';

describe('Menu Edit Page', () => {
  beforeEach(() => jest.useFakeTimers());
  afterEach(() => jest.useRealTimers());

  test('should render MenuEdit', () => {
    render(<MenuEdit menus={MOCK_MENU_ITEMS} />);
  });

  test('should render all menu names in a sorted way', () => {
    const sortedMenus = [...MOCK_MENU_ITEMS].sort((a, b) =>
      a.name.localeCompare(b.name),
    );
    render(<MenuEdit menus={MOCK_MENU_ITEMS} />);

    const allItemNames = screen.queryAllByText(/item/i);

    expect(allItemNames).toHaveLength(sortedMenus.length);
    sortedMenus.map((item, index) => {
      expect(allItemNames[index].props.children).toEqual(item.name);
    });
  });

  test('should render all checkboxes checked/unchecked state', () => {
    const checkedItems = MOCK_MENU_ITEMS.filter(item => item.isActive);
    const uncheckedItems = MOCK_MENU_ITEMS.filter(item => !item.isActive);
    render(<MenuEdit menus={MOCK_MENU_ITEMS} />);

    MenuEditTestHelpers.expectCheckedBoxes(checkedItems);
    MenuEditTestHelpers.expecUncheckedBoxes(uncheckedItems);
  });

  test('should mark as checked when press unchecked item', async () => {
    const checkedItems = MOCK_MENU_ITEMS.filter(item => item.isActive);
    const uncheckedItem = MOCK_MENU_ITEMS.filter(item => !item.isActive)[0];
    render(<MenuEdit menus={MOCK_MENU_ITEMS} />);

    MenuEditTestHelpers.expectCheckedBoxes(checkedItems);

    const unCheckedItemElement = screen.getByTestId(uncheckedItem.id);
    expect(unCheckedItemElement).toBeOnTheScreen();

    await userEvent.press(unCheckedItemElement);
    MenuEditTestHelpers.expectCheckedBoxes(
      checkedItems,
      checkedItems.length + 1,
    );
    const uncheckedItemChecbox = screen.getByTestId(
      `checked-${uncheckedItem.id}`,
    );
    expect(uncheckedItemChecbox).toBeOnTheScreen();
  });
});

class MenuEditTestHelpers {
  static expectCheckedBoxes = (
    checkedItems: IMenuItem[],
    expectedCount?: number,
  ) => {
    const checkedBoxes = screen.queryAllByText(/✅/i);
    expect(checkedBoxes).toHaveLength(expectedCount ?? checkedItems.length);
  };

  static expecUncheckedBoxes = (uncheckedItems: IMenuItem[]) => {
    const unCheckedBoxes = screen.queryAllByText(/❌/i);
    expect(unCheckedBoxes).toHaveLength(uncheckedItems.length);
  };
}

export const MOCK_MENU_ITEMS: IMenuItem[] = [
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe119999',
    name: 'a-Item 1',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe120121',
    name: 'f-Item 2',
    isActive: false,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe132131',
    name: 'e-Item 3',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe1191231',
    name: 'b-Item 4',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe141901',
    name: 'g-Item 5',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe119912',
    name: 'p-Item 6',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe14324',
    name: 'Item 7',
    isActive: true,
  },
  {
    id: '7a379770-0524-4ab2-91a9-fcd9fe11943',
    name: 'd-Item 8',
    isActive: false,
  },
];
