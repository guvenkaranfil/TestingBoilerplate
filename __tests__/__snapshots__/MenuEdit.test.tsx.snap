// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Edit Page should mark as checked when press unchecked item 1`] = `
[
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "❌",
  "❌",
]
`;

exports[`Menu Edit Page should mark as checked when press unchecked item 2`] = `
[
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "❌",
]
`;

exports[`Menu Edit Page should mark as unchecked when press checked item 1`] = `
[
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "❌",
  "❌",
]
`;

exports[`Menu Edit Page should mark as unchecked when press checked item 2`] = `
[
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "❌",
  "❌",
  "❌",
]
`;

exports[`Menu Edit Page should render all checkboxes checked/unchecked state 1`] = `
[
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "✅",
  "❌",
  "❌",
]
`;

exports[`Menu Edit Page should render all menu names in a sorted way 1`] = `
[
  "a-Item 1",
  "b-Item 4",
  "d-Item 8",
  "e-Item 3",
  "f-Item 2",
  "g-Item 5",
  "Item 7",
  "p-Item 6",
]
`;
